### PR TITLE
Avoid GPU syncs by reusing Pre-allocated Zero Tensor

### PR DESCRIPTION
**Pull request by quanta42**
This commit improves the FullyShardedDataParallel (FSDP) class in PyTorch by reducing unnecessary GPU synchronizations by reusing a pre-allocated zero tensor.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -522,6 +522,7 @@ def __init__(
         # implemented using post-save and pre-load hooks
         _init_state_dict_state(self)
         _register_all_state_dict_hooks(self)
+        self.zero = torch.tensor(0.0, device=self.compute_device)
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:525]: Should we initialize this lazily (upon calling `clip_grad_norm_` and actually needing to use this)?

Maybe we can put it in a private attribute like `self._zero_scalar`?

**Comment by awgu** [comment]
Thanks @quanta42 for the contribution! Could you sign the CLA?

**Comment by quanta42** [comment]
> Thanks @quanta42 for the contribution! Could you sign the CLA?

I did sign it but I work for Adobe so I need the authorization from the administrators.
I also cc them in slack, hopefully, they will get back to me soon.

Thanks for the prompt review!

**Review by awgu — APPROVED** [review]
LGTM! Could you fix lint?

**Comment by quanta42** [comment]
@awgu I was wondering why this hasn't been merged after fixing the linting?
I don't have the authorization to merge this and it seems that now tests are failing probably because of code update?
What's the process to merge this type of changes?

**Comment by awgu** [comment]
sorry this fell through the cracks
let me rebase and try to remerge

**Comment by quanta42** [comment]
> sorry this fell through the cracks let me rebase and try to remerge

Thank you

**Pull request closed without merging**